Add builds and tests with coverage

alexey-milovidov opened the pull request:
### Changelog category (leave one):
- Build/Testing/Packaging Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added builds and tests with coverage collection with introspection. Continuation of #56102.

alexey-milovidov committed:
Enable coverage for debug build

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #58792 from ClickHouse/ClickHouse:enable-coverage-for-debug-build -->
*This is an automated comment for commit 1afc5e8c01685d1bb3e86b5a0fff55618db517b0 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58792/1afc5e8c01685d1bb3e86b5a0fff55618db517b0/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58792/1afc5e8c01685d1bb3e86b5a0fff55618db517b0/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

robot-clickhouse committed:
Automatic style fix

alexey-milovidov committed:
Fix Python

alexey-milovidov committed:
Merge branch 'enable-coverage-for-debug-build' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Fix error

alexey-milovidov committed:
Slightly better

alexey-milovidov commented:
New debug tests: 1hr 23min .. 1hr 55min.

alexey-milovidov commented:
Old debug tests: 1hr 19min .. 1hr 23min.

Not a big difference - ok to merge.

alexey-milovidov committed:
Documentation

alexey-milovidov committed:
Report coverage if available

alexey-milovidov committed:
Calculate cumulative coverage by default.

alexey-milovidov committed:
Slightly better reporting

alexey-milovidov committed:
More functions

alexey-milovidov committed:
Fix typo

alexey-milovidov committed:
An option to dump coverage to a file at exit

alexey-milovidov committed:
Rename to system.coverage_log to simplify export

alexey-milovidov committed:
Export coverage to the CI database

alexey-milovidov commented:
As tests are run in parallel, so far, we will have only cumulative, not per-test coverage.

alexey-milovidov committed:
Symbolization

alexey-milovidov committed:
Symbolization

alexey-milovidov committed:
Fix error

alexey-milovidov committed:
Better dump on exit

alexey-milovidov committed:
Coverage for non-server tools

alexey-milovidov committed:
Obey Python's quirky formatter

alexey-milovidov committed:
Revert "Obey Python's quirky formatter"

This reverts commit 678a32cedee768b6c1a6748e96a0d103e853d8bc.

alexey-milovidov committed:
Obey Python's quirky formatter

alexey-milovidov committed:
Follow-up

alexey-milovidov committed:
Fix typo

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Maybe better

alexey-milovidov committed:
Better test

alexey-milovidov committed:
Remove obsolete comment

alexey-milovidov commented:
It started to become very slow after I added the coverage dump in command line tools.
Either I will introduce a separate release+coverage build or remove this dump.

alexey-milovidov committed:
Merge branch 'master' into enable-coverage-for-debug-build

alexey-milovidov committed:
Slightly better

alexey-milovidov committed:
Slightly better

alexey-milovidov committed:
Add a release build with coverage, turn off coverage in the debug build

alexey-milovidov committed:
Fix YAML

alexey-milovidov committed:
Fix error

alexey-milovidov committed:
Fix CMake

alexey-milovidov committed:
Add missing comments in code

robot-clickhouse committed:
Automatic style fix

alexey-milovidov committed:
Disable LTO with Coverage

alexey-milovidov committed:
Merge branch 'enable-coverage-for-debug-build' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Do not check for large translation units with coverage

alexey-milovidov committed:
Add tests with coverage

alexey-milovidov committed:
Ci to CI

alexey-milovidov committed:
Fix error

alexey-milovidov committed:
Fix error

alexey-milovidov committed:
Merge branch 'master' into enable-coverage-for-debug-build

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov commented:
Writing cumulative coverage was a bad idea - it accumulated 
```
total_rows:                    64525
total_bytes:                   455251337878
```
from just a single run.

alexey-milovidov committed:
Reset coverage after each test

alexey-milovidov commented:
It still generates too much data:
```
total_rows:                    5870
total_bytes:                   41913908675
```

Just from a single run.

alexey-milovidov commented:
Found a mistake.

alexey-milovidov committed:
Minor change

alexey-milovidov committed:
Fix error

alexey-milovidov commented:
Looks promising, maybe it has a chance.

alexey-milovidov committed:
Merge branch 'master' into enable-coverage-for-debug-build

alexey-milovidov committed:
Fix conflicts.

alexey-milovidov committed:
Split by batches

alexey-milovidov committed:
Fix Python

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov commented:
Not every test is reported in the coverage_log and it's strange that some irrelevant checks started to report coverage:

```
clickhouse-cloud :) SELECT check_start_time, check_name, count(), uniq(test_name), uniqArray(coverage) FROM coverage_log_11609173080111304637 GROUP BY ALL ORDER BY count() DESC LIMIT 10

SELECT
    check_start_time,
    check_name,
    count(),
    uniq(test_name),
    uniqArray(coverage)
FROM coverage_log_11609173080111304637
GROUP BY ALL
ORDER BY count() DESC
LIMIT 10

Query id: 3f8986f2-9de9-4c98-a413-ec36da9c646c

┌────check_start_time─┬─check_name────────────────────────────────────┬─count()─┬─uniq(test_name)─┬─uniqArray(coverage)─┐
│ 2024-01-29 18:23:45 │ Stateless tests (release, analyzer)           │    1241 │             134 │              227870 │
│ 2024-01-29 18:23:40 │ Stateless tests (release)                     │    1212 │             134 │              229618 │
│ 2024-01-29 18:24:01 │ Stateless tests (release, s3 storage)         │    1165 │             159 │              248876 │
│ 2024-01-29 18:24:07 │ Stateless tests (release, s3 storage)         │    1129 │             156 │              230304 │
│ 2024-01-21 15:40:45 │ Stateless tests (coverage)                    │    1088 │             229 │              255072 │
│ 2024-01-29 20:14:44 │ Stateless tests (coverage)                    │     447 │              48 │              178605 │
│ 2024-01-29 20:13:43 │ Stateless tests (coverage)                    │     431 │              39 │              174993 │
│ 2024-01-29 18:23:24 │ Stateless tests (release, DatabaseReplicated) │     424 │              50 │              194357 │
│ 2024-01-29 20:11:09 │ Stateless tests (coverage)                    │     407 │              60 │              185565 │
│ 2024-01-29 18:21:55 │ Stateless tests (release, DatabaseReplicated) │     402 │              60 │              193126 │
└─────────────────────┴───────────────────────────────────────────────┴─────────┴─────────────────┴─────────────────────┘
```

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Merge branch 'master' into enable-coverage-for-debug-build

alexey-milovidov committed:
Fix error

alexey-milovidov committed:
A few missing changes

robot-clickhouse committed:
Automatic style fix

alexey-milovidov committed:
Fix your Py

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Merge branch 'enable-coverage-for-debug-build' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Fix your Py

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Fix a test

Felixoid commented on tests/ci/ci_config.py:
```diff
@@ -713,6 +725,13 @@ def validate(self) -> None:
+            compiler="clang-17",
+            coverage=True,
+            package_type="deb",
+            sparse_checkout=True,
```
Please, don't use it. It slows down the checkout

Felixoid commented on .github/workflows/master.yml:
```diff
@@ -97,6 +97,14 @@ jobs:
       build_name: package_release
       checkout_depth: 0
       data: ${{ needs.RunConfig.outputs.data }}
+  BuilderDebReleaseCoverage:
```
You need to add BuilderDebReleaseCoverage to the BuilderReport job below. Here and in the pull_request.yml

alexey-milovidov committed:
Fix report

alexey-milovidov committed:
Reading Docker files

alexey-milovidov committed:
Better logs in CI and non-interactive install

alexey-milovidov committed:
English (somewhat incomplete)

alexey-milovidov committed:
Fix an error by rewriting the CI from Perl to Python

alexey-milovidov committed:
Better

alexey-milovidov committed:
Probably better

robot-clickhouse committed:
Automatic style fix

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov committed:
Run coverage collection sequentially

alexey-milovidov committed:
Merge branch 'enable-coverage-for-debug-build' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

Felixoid commented on tests/ci/ci_config.py:
```diff
@@ -199,13 +202,21 @@ class JobConfig:
     release_only: bool = False
 
 
+# About the "sparse_checkout" option:
+#
+# It's needed only to test the option itself.
+# No particular sense to use it in the build,
+# and it is even slows down the job.
+
+
 @dataclass
 class BuildConfig:
     name: str
     compiler: str
     package_type: Literal["deb", "binary", "fuzzers"]
     additional_pkgs: bool = False
     debug_build: bool = False
+    coverage: bool = False
     sanitizer: str = ""
     tidy: bool = False
```
Let's put it close to the field

```suggestion
@dataclass
class BuildConfig:
    name: str
    compiler: str
    package_type: Literal["deb", "binary", "fuzzers"]
    additional_pkgs: bool = False
    debug_build: bool = False
    coverage: bool = False
    sanitizer: str = ""
    tidy: bool = False
    # sparse_checkout is needed only to test the option itself.
    # No particular sense to use it in every build, since it slows down the job.
```

Felixoid commented on tests/queries/0_stateless/02982_comments_in_system_tables.sh:
```diff
@@ -4,5 +4,5 @@ CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 
-${CLICKHOUSE_LOCAL}  --query "SELECT 'Table ' || database || '.' || name || ' doesnt have a comment' FROM system.tables WHERE name NOT LIKE '%\_log\_%' AND database='system' AND comment==''"
-${CLICKHOUSE_CLIENT} --query "SELECT 'Table ' || database || '.' || name || ' doesnt have a comment' FROM system.tables WHERE name NOT LIKE '%\_log\_%' AND database='system' AND comment==''"
+${CLICKHOUSE_LOCAL}  --query "SELECT 'Table ' || database || '.' || name || ' doesn\'t have a comment' FROM system.tables WHERE name NOT LIKE '%\_log\_%' AND database='system' AND comment==''"
+${CLICKHOUSE_CLIENT} --query "SELECT 'Table ' || database || '.' || name || ' doesn\'t have a comment' FROM system.tables WHERE name NOT LIKE '%\_log\_%' AND database='system' AND comment==''"
```
would it be better to avoid escapes by using `does not`?

```suggestion
${CLICKHOUSE_LOCAL}  --query "SELECT 'Table ' || database || '.' || name || ' does not have a comment' FROM system.tables WHERE name NOT LIKE '%\_log\_%' AND database='system' AND comment==''"
${CLICKHOUSE_CLIENT} --query "SELECT 'Table ' || database || '.' || name || ' does not have a comment' FROM system.tables WHERE name NOT LIKE '%\_log\_%' AND database='system' AND comment==''"
```

alexey-milovidov committed:
Update tests/queries/0_stateless/02982_comments_in_system_tables.sh

Co-authored-by: Mikhail f. Shiryaev <felixoid@clickhouse.com>

alexey-milovidov committed:
Update tests/ci/ci_config.py

Co-authored-by: Mikhail f. Shiryaev <felixoid@clickhouse.com>

alexey-milovidov commented:
Almost ready, but:

1. It's strange that not every test is visible in the `coverage_log` on the CI Logs cluster.
2. I tried to make the tests sequential for the coverage runs, but it didn't work.
3. Check the hangs in some of the runs.

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov commented:
Nice: it captures almost all the tests:

```
SELECT
    check_start_time,
    check_name,
    uniq(test_name),
    sum(length(coverage))
FROM coverage_log_11609173080111304637
GROUP BY ALL
    WITH TOTALS
ORDER BY check_start_time ASC

Query id: a3812874-c04e-4246-863c-c26f1644d457

┌────check_start_time─┬─check_name─────────────────┬─uniq(test_name)─┬─sum(coverage.size0)─┐
│ 2024-02-09 23:17:00 │ Stateful tests (coverage)  │             134 │             9511806 │
│ 2024-02-09 23:17:04 │ Stateless tests (coverage) │             993 │           177304800 │
│ 2024-02-09 23:17:08 │ Stateless tests (coverage) │             724 │           132668437 │
│ 2024-02-09 23:17:24 │ Stateless tests (coverage) │             981 │           126618522 │
│ 2024-02-09 23:18:18 │ Stateless tests (coverage) │             733 │           138739289 │
│ 2024-02-09 23:19:46 │ Stateless tests (coverage) │            1027 │           183750228 │
│ 2024-02-09 23:20:20 │ Stateless tests (coverage) │             529 │            79867396 │
└─────────────────────┴────────────────────────────┴─────────────────┴─────────────────────┘

Totals:
┌────check_start_time─┬─check_name─┬─uniq(test_name)─┬─sum(coverage.size0)─┐
│ 1970-01-01 00:00:00 │            │            5115 │           848460478 │
└─────────────────────┴────────────┴─────────────────┴─────────────────────┘
```

alexey-milovidov committed:
Sequential run

alexey-milovidov commented:
> It's strange that not every test is visible in the coverage_log on the CI Logs cluster.

Fixed.

> I tried to make the tests sequential for the coverage runs, but it didn't work.

Fixed.

alexey-milovidov committed:
Fix "Too many parts"

alexey-milovidov committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into enable-coverage-for-debug-build

alexey-milovidov commented:
Finally, it looks alright.

alexey-milovidov commented:
Next step - add reports.

alexey-milovidov commented:
I suspect a subset of data is not delivered to the CI logs cluster due to high volume.
